File: typescript/src/sorting/insertion-sort.ts

/**
 * insertion sort implementation
 * it does sort the colleciton In Place
 * 
 * # example
 * ```ts
 * let arr = [1, 3, 2, 8, 4];
 * arr.sort();
 * ```
 * @param nums the array to sort
 * @returns sorted array
 */
export function insertionSort(nums: number[]) {
	for (let i = 0; i < nums.length ; i++){
		const cur = nums[i]
		let j = i;
		while (j > 0 && nums[j - 1] > cur) {
			nums[j] = nums[j - 1]; 
			j--;
		}
		nums[j] = cur;
	}
}

File: src/sorting/insertion_sort.c
#include "algorithms.h"

void insertion_sort(int arr[], int len) {
    int i = 0;
    int j = 0;
    int key = 0;
    for (i = 1; i < len; i++) {
        key = arr[i];
        j = i - 1;
        while (j >= 0 && arr[j] > key) {
            arr[j + 1] = arr[j];
            j = j - 1;
        }
        arr[j + 1] = key;
    }
}
